### PR TITLE
Core/hubot

### PR DESCRIPTION
**Pull request by mrkno**

### CODE REVIEW

File: core/core_modules/issue.js [hunk 1]
```diff
@@ -111,22 +110,23 @@ ifStringNotEmpty = function(str) {
...
+    return [
+        [commandPrefix + 'issue "<title>" <debugLevel>', 'Posts an issue to Github.', 'Posts an issue to Github. DebugLevel can be basic, detail or full, defaults to basic if notspecified.'],
+        [commandPrefix + 'issue "<title>" "<description>" <debugLevel>', 'Posts an issue to Github with a description.', 'Posts an issue to Github with a description. DebugLevel can be basic, detail or full, defaults to basic if not specified.']
+    ];
 };
 
-exports.match = function(text, commandPrefix) {
-    return text.startsWith(commandPrefix + command);
+exports.match = function(event, commandPrefix) {
+    return event.arguments[0] === commandPrefix + 'issue';
```
Comment (mrkno) [inline core/core_modules/issue.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645506/issues/source?bid=3282987&fileBranchId=3282997#l120)

File: core/shim.js [hunk 2]
```diff
@@ -68,20 +74,23 @@ exports.createPlatformModule = function(platform) {
...
+	var event = {
 		thread_id: thread,
 		sender_id: senderId,
 		sender_name: senderName,
 		body: message
-	};
+    };
+    event.arguments = event.body.match(/(?:[^\s"]+|"[^"]*")+/g);
+    if (event.arguments === null) {
+        event.arguments = [''];
```
Comment (mrkno) [inline core/shim.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645531/issues/source?bid=3282987&fileBranchId=3282997#l89)

File: core/core_modules/issue.js [hunk 3]
```diff
@@ -9,8 +9,7 @@
  *		Copyright (c) Dion Woolley and Contributors 2016.
  */
 
-var command = "issue",
-    githubAPI = "https://api.github.com/repos/mrkno/Kassy/issues",
+var githubAPI = "https://api.github.com/repos/mrkno/Kassy/issues",
     token = 'MjBlNGU3YzM0Mjk5MDI2MzI1Y2M4MDUyODBjZGE4YzAwYmQ0YjYzZQ==',
     request = require.safe('request'),
```
Comment (mrkno) [inline core/core_modules/issue.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645506/issues/source?bid=3282987&fileBranchId=3282997#l14)

File: core/modules/hubotModule.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ * License:
+ *		MIT License. All code unless otherwise specified is
+ *		Copyright (c) Matthew Knox and Contributors 2015.
+ */
+
+var fs = require('fs'),
+    path = require('path'),
+    files = require.once('./../files.js'),
+    config = require('./../config.js'),
+    modulesDir = 'modules',
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l16)

File: core/modules/hubotModule.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+exports.verifyModule = function (location, disabled) {
+    var stat = fs.statSync(location);
+    if (!stat.isDirectory()) {
+        return null;
+    }
+    
+    var folderPath = path.resolve(location),
+        desc = path.join(folderPath, './' + descriptor),
+        pack = path.join(folderPath, './' + pkg),
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l42)

File: core/modules/hubotModule.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        hj;
+
+    try {
+        fs.statSync(desc);
+        hj = require.once(desc);
+    }
+    catch (e) {
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Closing curly brace does not appear on the same line as the subsequent block.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l49)

File: core/modules/hubotModule.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    
+    for (var i = 0; i < data.length; i++) {
+        try {
+            var candidate = path.resolve(path.join(modulesDir, data[i])),
+                output = exports.verifyModule(candidate, disabled);
+            if (output) {
+                modules[output.name] = output;
+            }
+            else {
+                console.debug('Skipping "' + data[i] + '". It isn\'t a Hubot module.');
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Unexpected console statement.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l91)

File: core/modules/hubotModule.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+                l.push(module.help[i][j].replace(/{{commandPrefix}}/g, commandPrefix));
+            }
+            h.push(l);
+        }
+        return h;
+    };
+};
+
+exports.loadModule = function (module) {
+    if (!coffeescriptLoaded && module.startup.endsWith('.coffee')) {
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l117)

File: core/modules/hubotModule.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        var modulePath = module.folderPath,
+            startPath = path.join(modulePath, module.startup),
+            m = require.once(startPath);
+
+        var cfg = config.loadModuleConfig(module, modulePath);
+        return new Robot(m, module, cfg);
+    }
+    catch (e) {
+        console.critical(e);
+        throw 'Could not load module \'' + module.name + '\'. Does it have a syntax error?';
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l132)

File: core/modules/hubotModule.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ * Written By:
+ * 		Matthew Knox
+ *
+ * License:
+ *		MIT License. All code unless otherwise specified is
+ *		Copyright (c) Matthew Knox and Contributors 2015.
+ */
+
+var fs = require('fs'),
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use doublequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2529645521/issues/source?bid=3282987&fileBranchId=3282997#l12)

File: core/modules/hubot/message.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+    this.room = event.thread_id;
+    this.text = event.body;
+    this.user = {
+        name: event.sender_name,
+        id: event.sender_id,
+        email_address: 'kassy@foobar.com'
+    };
+};
+
+module.exports = Message;
```
Comment (mrkno) [inline core/modules/hubot/message.js]: Codacy Issue found: [Newline required at end of file but not found.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065977/issues/source?bid=3282987&fileBranchId=3282997#l14)

File: core/modules/hubot/message.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿var Message = function (event, prefix) {
```
Comment (mrkno) [inline core/modules/hubot/message.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065977/issues/source?bid=3282987&fileBranchId=3282997#l1)

File: core/modules/hubotModule.js [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        } catch (e) {
+            console.debug('A failure occured while listing "' + data[i] + '". It doesn\'t appear to be a module.');
+            console.critical(e);
+            continue;
+        }
+    }
+    return modules;
+};
+
+var createHelp = function (module) {
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: ['createHelp' is defined but never used.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l102)

File: core/modules/hubotModule.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        desc = path.join(folderPath, './' + descriptor),
+        pack = path.join(folderPath, './' + pkg),
+        hj;
+
+    try {
+        fs.statSync(desc);
+        hj = require.once(desc);
+    }
+    catch (e) {
+        var l;
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: ['l' is defined but never used.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l50)

File: core/modules/hubotModule.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            h.push(l);
+        }
+        return h;
+    };
+};
+
+exports.loadModule = function (module) {
+    if (!coffeescriptLoaded && module.startup.endsWith('.coffee')) {
+        require("coffee-script").register();
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l118)

File: core/modules/hubotModule.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    Robot = require.once('./hubot/robot.js'),
+    coffeescriptLoaded = false;
+
+var verifyModuleDescriptior = function (hj, disabled) {
+    if (!hj.name || !hj.startup || !hj.version) {
+        return false;
+    }
+    
+    if (disabled === true && exports.disabledConfig 
+        && exports.disabledConfig[hj.name] && exports.disabledConfig[hj.name] === true) {
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: ['&&' should be placed at the end of the line.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l28)

File: core/modules/hubotModule.js [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+    pkg = 'package.json',
+    Robot = require.once('./hubot/robot.js'),
+    coffeescriptLoaded = false;
+
+var verifyModuleDescriptior = function (hj, disabled) {
+    if (!hj.name || !hj.startup || !hj.version) {
+        return false;
+    }
+    
+    if (disabled === true && exports.disabledConfig
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l27)

File: core/modules/hubotModule.js [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+    descriptor = 'hubot.json',
+    pkg = 'package.json',
+    Robot = require.once('./hubot/robot.js'),
+    coffeescriptLoaded = false;
+
+var verifyModuleDescriptior = function (hj, disabled) {
+    if (!hj.name || !hj.startup || !hj.version) {
+        return false;
+    }
+
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Trailing spaces not allowed.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l26)

File: core/modules/hubotModule.js [hunk 19]
```diff
@@ -0,0 +1,134 @@
+﻿/**
```
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l1)

File: core/modules/hubotModule.js (hunk 15 above)
Comment (mrkno) [inline core/modules/hubotModule.js]: Codacy Issue found: [Strings must use singlequote.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530065980/issues/source?bid=3282987&fileBranchId=3282997#l118)

File: core/modules/hubot/message.js (hunk 11 above)
Comment (mrkno) [inline core/modules/hubot/message.js]: Codacy Issue found: [Newline required at end of file but not found.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104136/issues/source?bid=3282987&fileBranchId=3282997#l14)

File: core/modules/hubot/message.js (hunk 12 above)
Comment (mrkno) [inline core/modules/hubot/message.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104136/issues/source?bid=3282987&fileBranchId=3282997#l1)

File: core/install.js [hunk 20]
```diff
@@ -23,19 +24,25 @@ install = function(name) {
...
+		return req(name);
 	}
 	catch(e) {
 		if (!e || !e.code || e.code !== 'MODULE_NOT_FOUND') {
 			throw e;
 		}
+		try {
+			fs.statSync(name);
+		}
+		catch(p) {
```
Comment (mrkno) [inline core/install.js]: Codacy Issue found: [Keyword "catch" must be followed by whitespace.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104133/issues/source?bid=3282987&fileBranchId=3282997#l38)

File: core/platform.js [hunk 21]
```diff
@@ -10,8 +10,7 @@
 *        Copyright (c) Matthew Knox and Contributors 2015.
 */
 
-var figlet = require('figlet'),
-    shim = require.once('./shim.js'),
+var figlet = require('figlet');
 
 Platform = function(modes) {
```
Comment (mrkno) [inline core/platform.js]: Codacy Issue found: ["Platform" is not defined.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104146/issues/source?bid=3282987&fileBranchId=3282997#l15)

File: core/platform.js [hunk 22]
```diff
@@ -191,24 +188,18 @@ Platform.prototype.shutdown = function(flag) {
...
-        }
-        this.coreModules[i] = null;
+    while (this.coreModules.length > 0) {
+        this.coreLoader.unloadCoreModule(this.coreModules[0]);
+        this.coreModules.splice(0, 1);
     }
-    this.coreModules = [];
 
-    this.config.saveConfig();
+    this.config.saveSystemConfig();
```
Comment (mrkno) [inline core/platform.js]: Codacy Issue found: ["StatusFlag" is not defined.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104146/issues/source?bid=3282987&fileBranchId=3282997#l203)

File: core/modules/hubot/robot.js [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+            scriptDocumentation[currentSection].push(cleanedLine.trim());
+            if (currentSection === 'commands') {
+                commands.push(cleanedLine.trim());
+            }
+        }
+    }
+
+    var help = [];
+    if (scriptDocumentation.commands) {
+        for (var i = 0; i < scriptDocumentation.commands.length; i++) {
```
Comment (mrkno) [inline core/modules/hubot/robot.js]: Codacy Issue found: [Variable 'i' is already declared in line 48.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104138/issues/source?bid=3282987&fileBranchId=3282997#l72)

File: core/modules/hubot/robot.js [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            event.__robotCallbackListeners.push({
+                callback: this.listeners[i].callback,
+                match: m
+            });
+        }
+    }
+
+    if (!hasMatched) {
+        for (var i = 0; i < this.catchAllListeners.length; i++) {
```
Comment (mrkno) [inline core/modules/hubot/robot.js]: Codacy Issue found: [Variable 'i' is already declared in line 121.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104138/issues/source?bid=3282987&fileBranchId=3282997#l137)

File: core/modules/hubot/robot.js [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+        },
+        callback: callback
+    });
+};
+
+Robot.prototype.respond = function (regex, callback) {
+    this.listeners.push({
+        matcher: function (msg) {
+            var prefix = msg.prefix.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
+            var reg = eval('/^' + prefix + regex.toString().substring(1));
```
Comment (mrkno) [inline core/modules/hubot/robot.js]: Codacy Issue found: [Eval statement detected.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104138/issues/source?bid=3282987&fileBranchId=3282997#l172)

File: core/modules/hubot/robot.js [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+            'authors',
+            'examples',
+            'tags',
+            'urls'
+        ],
+        mod = path.join(folderPath, scriptLocation),
+        body = fs.readFileSync(mod, 'utf-8'),
+        scriptDocumentation = { name: path.basename(mod).replace(/\.(coffee|js)$/, '') },
+        currentSection = null,
+        ref = body.split("\n"),
```
Comment (mrkno) [inline core/modules/hubot/robot.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530104138/issues/source?bid=3282987&fileBranchId=3282997#l45)

File: core/core_modules/help.js [hunk 27]
```diff
@@ -1,8 +1,10 @@
 var figlet = require.safe('figlet'),
 
-constructHelpMessage = function(help, modules, context, prefix) {
-	for (var i = 0; i < modules.length; i++) {
-		var cmdHelp = modules[i].help.call(context, prefix);
+constructHelpMessage = function (help, modules, context) {
+    for (var i = 0; i < modules.length; i++) {
+        var cmdHelp = modules[i].ignoreHelpContext
+            ? modules[i].help(context.commandPrefix)
+            : modules[i].help.call(context, context.commandPrefix);
```
Comment (mrkno) [inline core/core_modules/help.js]: Codacy Issue found: [':' should be placed at the end of the line.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333283/issues/source?bid=3282987&fileBranchId=3282997#l7)

File: core/core_modules/help.js [hunk 28]
```diff
@@ -11,47 +13,41 @@ constructHelpMessage = function(help, modules, context, prefix) {
...
 
 	var help = '',
-		context = {
-			commandPrefix: prefix
-		},
-		cmdHelp = module.help.call(context, prefix);
-		
+        cmdHelp = module.ignoreHelpContext
+            ? module.help(context.commandPrefix)
+            : module.help.call(context, context.commandPrefix);
```
Comment (mrkno) [inline core/core_modules/help.js]: Codacy Issue found: [':' should be placed at the end of the line.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333283/issues/source?bid=3282987&fileBranchId=3282997#l49)

File: core/core_modules/help.js [hunk 29]
```diff
@@ -1,8 +1,10 @@
 var figlet = require.safe('figlet'),
 
-constructHelpMessage = function(help, modules, context, prefix) {
-	for (var i = 0; i < modules.length; i++) {
-		var cmdHelp = modules[i].help.call(context, prefix);
+constructHelpMessage = function (help, modules, context) {
+    for (var i = 0; i < modules.length; i++) {
+        var cmdHelp = modules[i].ignoreHelpContext
+            ? modules[i].help(context.commandPrefix)
```
Comment (mrkno) [inline core/core_modules/help.js]: Codacy Issue found: ['?' should be placed at the end of the line.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333283/issues/source?bid=3282987&fileBranchId=3282997#l6)

File: core/platform.js (hunk 21 above)
Comment (mrkno) [inline core/platform.js]: Codacy Issue found: ["Platform" is not defined.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333306/issues/source?bid=3282987&fileBranchId=3282997#l15)

File: core/platform.js (hunk 22 above)
Comment (mrkno) [inline core/platform.js]: Codacy Issue found: ["StatusFlag" is not defined.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333306/issues/source?bid=3282987&fileBranchId=3282997#l203)

File: core/require.js [hunk 30]
```diff
@@ -13,12 +13,21 @@
...
+        return inst.requireOrInstall(req, mod);
+    };
+    for (var key in req) {
+        func[key] = req[key];
+    }
+    func.safe = func;
+
+    func.searchCache = function (moduleName, callback) {
+        var mod = func.resolve(moduleName);
+        if (mod && ((mod = func.cache[mod]) !== undefined)) {
```
Comment (mrkno) [inline core/require.js]: Codacy Issue found: [Unexpected use of undefined.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333308/issues/source?bid=3282987&fileBranchId=3282997#l30)

File: core/modules/hubot/responder.js [hunk 31]
```diff
@@ -0,0 +1,24 @@
...
+    for (var i = 0; i < arguments.length; i++) {
+        this.api.sendMessage(arguments[i], this.event.thread_id);
+    }
+};
+
+Responder.prototype.emote = Responder.prototype.send;
+Responder.prototype.reply = Responder.prototype.send;
+Responder.prototype.http = http.create;
+
+module.exports = Responder;
```
Comment (mrkno) [inline core/modules/hubot/responder.js]: Codacy Issue found: [Newline required at end of file but not found.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333297/issues/source?bid=3282987&fileBranchId=3282997#l24)

File: core/modules/hubot/responder.js [hunk 32]
```diff
@@ -0,0 +1,24 @@
+﻿var http = require('scoped-http-client');
```
Comment (mrkno) [inline core/modules/hubot/responder.js]: Codacy Issue found: [Irregular whitespace not allowed](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333297/issues/source?bid=3282987&fileBranchId=3282997#l1)

File: core/modules/hubot/robot.js [hunk 33]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            event.__robotCallbackListeners.push({
+                callback: this.listeners[i].callback,
+                match: m
+            });
+        }
+    }
+
+    if (!hasMatched) {
+        for (var i = 0; i < this.catchAllListeners.length; i++) {
```
Comment (mrkno) [inline core/modules/hubot/robot.js]: Codacy Issue found: [Variable 'i' is already declared in line 121.](https://www.codacy.com/app/theprojectshop/Kassy/file/2530333298/issues/source?bid=3282987&fileBranchId=3282997#l137)

**Comment by Awarua-** [comment]
Approved

**Pull request merged**